grin/src/miner.rs

```diff
@@ -527,22 +621,182 @@ impl Miner {
+							// There is a new block on the chain
+							// OR we are rebuilding the current one to include new transactions
+							if current_hash != latest_hash {
+								// A brand new block, so we will generate a new key_id
+								key_id = None;
+							}
+
+							let mut result = self.build_block(&head, key_id.clone());
+							while let Err(e) = result {
+								match e {
+									self::Error::Chain(chain::Error::DuplicateCommitment(_)) => {
+										debug!(LOGGER,
+										       "Duplicate commit for potential coinbase detected. Trying next derivation.");
+									}
+									ae => {
+										warn!(
+											LOGGER,
+											"Error building new block: {:?}. Retrying.", ae
+										);
+									}
+								}
+								thread::sleep(Duration::from_millis(100));
+								result = self.build_block(&head, key_id.clone());
+							} // error building block
+
+							let (new_block, block_fees) = result.unwrap();
+							b = new_block;
+							key_id = block_fees.key_id();
+							current_hash = latest_hash;
+							// set a new deadline for rebuilding with fresh transactions
+							deadline = time::get_time().sec + attempt_time_per_block as i64;
+
+							debug!(
+								LOGGER,
+								"(Server ID: {}) sending block {} to stratum client",
+								self.debug_output_id,
+								b.header.height
+							);
+
+							// Push the new block "b" to the connected client
+							self.stratum_send_block(&mut stream, &b, &mut stream_err);
+						}
+
+						// Get a solved block header - if any are waiting
+						// Here, we are expecting real solutions at the full difficulty.
+						match self.stratum_get_solution(&mut stream, &mut stream_err) {
+							Some(a_block_header) => {
+								// Got a solution
+								if a_block_header.height == b.header.height {
+									b.header = a_block_header;
+									info!(LOGGER,
+									      "(Server ID: {}) Found valid proof of work, adding block {}",
+									      self.debug_output_id,
+									      b.hash());
+									let res = self.chain.process_block(b.clone(), chain::NONE);
+									if let Err(e) = res {
+										error!(
+											LOGGER,
+											"(Server ID: {}) Error validating mined block: {:?}",
+											self.debug_output_id,
+											e
+										);
+									}
+									debug!(LOGGER, "resetting key_id in miner to None");
+								} else {
+									warn!(LOGGER,
+									      "(Server ID: {}) Found POW for block at height: {} -  but too late",
+									      self.debug_output_id,
+									      b.header.height);
+								}
+							}
+							None => {} /* No solutions found yet, let the client keep mining the
+							            * same block */
+						} // checking for solution
+		// sleep before restarting loop
+						thread::sleep(Duration::from_millis(500));
+					} // loop forever
+				} // end OK() match stream
+			} // match stream
+		} // for stream incoming
+	} // fn run_stratum_server_loop()
+
```

Comment:
Would try to split up the functionality here a bit into separate functions, but understand it's still evolving.